ethereum/trie/src/main/java/tech/pegasys/pantheon/ethereum/trie/SimpleMerklePatriciaTrie.java

```diff
@@ -51,6 +53,17 @@ public SimpleMerklePatriciaTrie(final Function<V, BytesValue> valueSerializer) {
     return root.accept(getVisitor, bytesToPath(key)).getValue();
   }
 
+  @Override
+  public Proof<V> getValueWithProof(final K key) {
+    checkNotNull(key);
+    final ProofVisitor<V> proofVisitor = new ProofVisitor<>();
+    final Optional<V> value = get(key);
```

Comment:
Instead of traversing twice, can we update the `ProofVisitor` to collect the `Optional<V> value` as it traverses?

Reply:
you are right. ProofVisitor inherits from GetVisitor so it is already possible to recover the value without traversing twice